README
This project provide tools to manipulate firmware for Built-in iSight
found on Apple machine since iMac G5 iSight (November 2005).

Those tools are based on the work from Ronald S. Bultje and are
licensed under the term of the GPLv2 or later. There are currently
(2010) maintained by Étienne Bersac.

See https://launchpad.net/isight-firmware-tools/ for bugs, code and
more informations.

ift-extract :
—————————————

	ift-extract extracts isight.fw from any AppleUSBVideoSupport
	driver from Darwin. This driver is located in
	/System/Library/Extensions/IOUSBFamily.kext/Contents/PlugIns/AppleUSBVideoSupport.kext/Contents/MacOS/AppleUSBVideoSupport
	on your Mac OS X root partition.

ift-load :
——————————

	ift-load load a firmware extracted by ift-extract right into
	the iSight. This tool is meant to be called by udev. (see
	isight.rules udev rules file provided).

ift-export :
——————————

	Exports a firmware extracted by ift-extract as Intel HEX
	format for load with fxload. Note that the RAM size of the
	Built-in iSight seems larger than 8KB, therefore fxload refuse
	to load the end of the firmware into the chip.


NEWS
Versio 1.6 :
------------

 * Now seek for firmware instead of using SHA1 sums + offset.

Version 1.5.90 :
----------------

 * Dropped HAL callout

 * Added SHA1 for Mac OS X.5.7, X.5.8 and X.6 driver on various 
   hardware.

 * Updated japanese translation (Thanks to Nobuhiro Iwamatsu)

Version 1.4.1 :
–––––––––––––––

 * Added three SHA1 sum for Mac OS X.6.

 * Order SHA1 sums to optimize lookup.

Version 1.4 :
–––––––––––––

 * Search vendor/product id rather than bus/dev id in udev loader.

 * Switch back to udev loader by default. This is less portable, but
   HAL and DeviceKit people prefer this policy.

 * Added two SHA1 sum from debian package and for Mac OS X.5.5

Version 1.2 :
—————————————

 * Provide HAL callout, udev support is obsolete. Use --enable-udev
   and --disable-hal to have previous behaviour.

 * Updated driver table.

Version 1.0 :
—————————————

 * Use DEVICE env var for finding USB bus id and device id.

 * Added Leopard driver support.

Version 0.5 :
—————————————

 * new tool : ift-export. Exports firmware in Intel HEX format.

 * Uses syslog facility in ift-load.

 * Info and man pages for ift-extract and ift-export.

Version 0.4 :
—————————————

 * Added i18n and fr_FR translation.

 * Use GOption parsing.

 * ift-extract: added option for output directory and output
   filename. Default to /lib/firmware and isight.fw.

 * ift-load: drop vendor_id:product_id probe and use udev USB_BUS and
   USB_DEV environment variables or from command line option.

Version 0.3 :
—————————————

 * added udev rules for automatic firmware loading.

 * install ift-load in sbin dir.

 * ift-extract now patch the firmware.


Version 0.2 :
—————————————

 * Splitted extract into ift-extract and ift-load.

 * Use autotools.

 * Added Mac OS X.4 PPC driver offset.


src/load.c
/*
 * Apple built-in iSight firmware loader
 *
 * Copyright © 2006 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright © 2007 Étienne Bersac <bersace@gmail.com>
 *
 *
 * Special thanks to Johannes Berg <johannes@sipsolutions.net> for helping
 * to find out how to load the firmware; see his website on
 * http://johannes.sipsolutions.net/MacBook/iSight for details.
 * Basic structure of firmware data: <len:2><off:2><data:len>, where
 * the $data of size $len is to be put at position $off in the device.
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA
 */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <syslog.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gstdio.h>


#include "load.h"
#define TIMEOUT 300

int
load_firmware(struct usb_device *dev, char *firmware)
{
	int fd, len, req, llen, res, ret = -1;
	unsigned char data[4], rdata[1024], *ptr;
	usb_dev_handle *h;

	/* load firmware */
	if (!(h = usb_open(dev))) {
		ift_warning(_("Failed to open device"));
		return -1;
	}

	if ((fd = g_open(firmware, O_RDONLY)) == -1) {
		ift_warning(_("Failed to open firmware"));
		return -1;
	}

	if ((res = usb_control_msg(h, 0x40, 0xA0, 0xe600, 0,
				   "\1", 1, TIMEOUT)) != 1) {
		ift_warning(_("Failed to init firmware loading"));
		close(fd);
		return -1;
	}
	while (1) {
		/* read length and offset */
		if ((len = read(fd, data, 4)) != 4) {
			if (len == 0) {
				break;
				goto end;
			} else {
				ift_warning(_("Reading firmware header chunk "
					      "failed"));
				goto end;
			}
		}

		len =(data[0] << 8) | data[1];
		req = (data[2] << 8) | data[3];

		/* read data */
		if (len < 0 || len >= 1024) {
			ift_warning(_("Invalid firmware data_length %d, "
				      "load aborted\n"),
				    len);
			goto end;
		} else if (read(fd, rdata, len) != len) {
			ift_warning(_("Failed to read firmware data"));
			goto end;
		}

		/* upload to usb bus */
		for (ptr = rdata; len > 0; req += 50, ptr += 50) {
			llen = len > 50 ? 50 : len;
			len -= llen;

			if ((res = usb_control_msg(h, 0x40, 0xA0, req, 0,
						   (char *) ptr, llen,
						   TIMEOUT)) != llen) {
				ift_warning(_("Firmware load req=0x%x failed: %s"),
					    req, strerror(errno));
				goto end;
			}
		}
	}

	ret = 0;

 end:
	if ((res = usb_control_msg(h, 0x40, 0xA0, 0xe600, 0,
				   "\0", 1, TIMEOUT)) != 1) {
		ift_warning(_("Failed to close firmware loading"));
		ret = -1;
	}

	close(fd);
	usb_close(h);

	return ret;
}

struct usb_device*
find_usb_device(const gchar *bus_id, const gchar *dev_id)
{
	struct usb_bus *bi, *bus = NULL;
	struct usb_device *di, *dev = NULL;

	for (bi = usb_busses; bi != NULL; bi = bi->next) {
		if (g_str_equal(bi->dirname, bus_id)) {
			bus = bi;
			break;
		}
	}

	if (!bus) {
		ift_warning(_("USB Bus %s not found"), bus_id);
		return NULL;
	}

	for (di = bus->devices; di != NULL; di = di->next) {
		if (g_str_equal(di->filename, dev_id)) {
			dev = di;
			break;
		}
	}

	if (!dev)
		ift_warning(_("USB device %s not found in bus %s"),
			    dev_id, bus_id);

	return dev;
}

struct usb_device*
find_usb_product(const guint32 vendor_id, const guint32 product_id)
{
	struct usb_bus *bi;
	struct usb_device *di, *dev = NULL;

	for (bi = usb_busses; bi != NULL; bi = bi->next) {
		for (di = bi->devices; di != NULL; di = di->next) {
			if (di->descriptor.idVendor == vendor_id &&
			    di->descriptor.idProduct == product_id) {
				dev = di;
				break;
			}
		}

		if (dev)
			break;
	}

	if (!dev)
		ift_warning(_("USB device 0x%04X:0x%04X not found"),
			    vendor_id, product_id);

	return dev;
}


src/load.h
/*
 * Apple built-in iSight firmware loader
 *
 * Copyright © 2006 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright © 2007 Étienne Bersac <bersace@gmail.com>
 *
 *
 * Special thanks to Johannes Berg <johannes@sipsolutions.net> for helping
 * to find out how to load the firmware; see his website on
 * http://johannes.sipsolutions.net/MacBook/iSight for details.
 * Basic structure of firmware data: <len:2><off:2><data:len>, where
 * the $data of size $len is to be put at position $off in the device.
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA
 */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <glib.h>
#include <usb.h>

#ifndef _LOAD_H_
#define	_LOAD_H_

#define APPLE_VENDOR_ID		0x05AC
#define	ISIGHT_PRODUCT_ID	0x8300

#define	ift_debug(args...)	syslog(LOG_DEBUG, args)
#define	ift_message(args...)	syslog(LOG_INFO, args)
#define	ift_warning(args...)	syslog(LOG_WARNING, args)
#define	ift_error(args...)	({syslog(LOG_ERR, args); exit(-1);})

int			load_firmware(struct usb_device *dev, char *firmware);
struct usb_device*	find_usb_device(const gchar *bus_id, const gchar *dev_id);
struct usb_device*	find_usb_product(const guint32 vendor_id, const guint32 product_id);

#endif


src/udev.c
/*
 * Apple built-in iSight firmware loader
 *
 * Copyright © 2006 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright © 2007 Étienne Bersac <bersace@gmail.com>
 *
 *
 * Special thanks to Johannes Berg <johannes@sipsolutions.net> for helping
 * to find out how to load the firmware; see his website on
 * http://johannes.sipsolutions.net/MacBook/iSight for details.
 * Basic structure of firmware data: <len:2><off:2><data:len>, where
 * the $data of size $len is to be put at position $off in the device.
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA
 */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <syslog.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gstdio.h>

#include "load.h"

/* OPTIONS */
static gchar		*firmware	= "isight.fw";

static GOptionEntry entries[] = {
	{ "firmware", 'f',
	  0,
	  G_OPTION_ARG_FILENAME, &firmware,
	  N_("Path to the firmware."),
	  NULL
	},
	{ NULL }
};

int
main(int argc, char *argv[])
{
	GOptionContext *context;
	GError *error = NULL;
	struct usb_device *dev;

	openlog("ift-load", LOG_PERROR, LOG_USER);

#if ENABLE_NLS
	/* ??????? */
	setlocale(LC_ALL, "");
	bindtextdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);
#endif

	gchar *appname = _("built-in iSight firmware loader");

	g_set_application_name(appname);
	g_set_prgname("ift-load");

	/* translators: this is the first --help output line */
	context = g_option_context_new(g_strdup_printf(" - %s",
						       appname));

	/* tip at the end of --help output */
	gchar *desc = _("Use ift-extract to extract firmware.");
	g_option_context_set_description(context, desc);
	/* translators: this is shown at the end of --help output */
	g_option_context_add_main_entries(context, entries, GETTEXT_PACKAGE);
	if (!g_option_context_parse(context, &argc, &argv, &error)) {
		fprintf(stderr, _("Error: %s\n"),
			error->message); /* comment traduire ? */
		fprintf(stderr, g_option_context_get_help(context,
							  FALSE, NULL));
	}

	if (g_access(firmware, R_OK)) {
		ift_error(_("Unable to read firmware %s."), firmware);
	}

	/* init usb */
	usb_init();
	if (usb_find_busses() == 0) {
		ift_error(_("No USB busses found"));
	} else if (usb_find_devices() == 0) {
		ift_error(_("No USB devices found"));
	}


	dev = find_usb_product(APPLE_VENDOR_ID,
			       ISIGHT_PRODUCT_ID);

	if (!dev)
		ift_error(_("No iSight found"));

	if (load_firmware(dev, firmware) == -1)
		ift_error(_("Failed to upload firmware to 0x%04X:0x%04X"),
			  dev->descriptor.idVendor,
			  dev->descriptor.idProduct);
	else
		ift_message(_("iSight firmware loaded successfully"));

	closelog();

	return 0;
}


src/export.c
/*
 * Apple built-in iSight firmware exporter
 *
 * Copyright © 2006 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright © 2007 Étienne Bersac <bersace@gmail.com>
 *
 *
 * This software export Mac OS X firmware to Intel HEX format.
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gstdio.h>

/* OPTIONS */
static gchar *input	= NULL;
static gchar *output	= "isight.ihx";

static GOptionEntry entries[] = {
	{ "input-firmware", 'i',
	  0,
	  G_OPTION_ARG_FILENAME, &input,
	  N_("Path to the extracted firmware."),
	  NULL
	},
	{ "output", 'o',
	  0,
	  G_OPTION_ARG_FILENAME, &output,
	  N_("Output Intel HEX filename"),
	  "isight.ihx"
	},
	{ NULL }
};


/* IMPLEMENTATION */

/* See http://en.wikipedia.org/wiki/Intel_HEX */
static guchar hex_checksum(guchar len, guint16 add, guchar type, guchar *data)
{
	gint total;
	guchar i, res, checksum;

	total = len + add + type;	
	for (i = 0; i < len; i++) {
		total+=data[i];
	}
	res = total & 0x000000ff;
	checksum = (0x100 - res) & 0xff;

	return checksum;
}

/* Write a line to fd following Intel HEX format */
static int write_line(int fd,
		      guchar len, guint16 add, guchar type, guchar *data)
{
	gchar *line, *sdata = g_malloc0((2*len+1) * sizeof(guchar));
	guchar checksum = hex_checksum(len, add, type, data);
	guchar i, w;

	for (i = 0; i < len; i++) {
		sprintf(sdata+2*i, "%02X", data[i]);
	}

	line = g_strdup_printf(":%02X%04X%02X%s%02X\n",
			       len, add, type, sdata, checksum);
	w = write(fd, line, strlen(line));

	if (w != strlen(line)) {
		g_error(_("Error while writing '%s', "
			  "%i bytes written instead of %i"),
			line, w, strlen(line));
	}
	g_free(line);
	g_free(sdata);

	return 0;
}

/* Split a chunk of data in Intel HEX line, output to fd  */
static int write_chunk(int fd, gint len, gint add, guchar *data)
{
	gint i;
	guchar type = 0x00;
	guchar dlen;

	for (i = 0; i < len; i+= 32) {
		dlen = MIN(32, len - i);
		write_line(fd, dlen, (gint16) (add + i), type, data + i);
	}

	return 0;
}

/* extract firmware from filename at offset to output, checking the
   format. */
static int
export(char *filename,
       char*output)
{
	int fd, fdo, len, req, ret = -1;
	unsigned char data[4], rdata[1024];

	if ((fd = open (filename, O_RDONLY)) == -1) {
		g_error(_("Unable to open %s."), filename);
		return -1;
	}

	if ((fdo = open(output, O_WRONLY | O_CREAT | O_TRUNC,
			S_IRUSR | S_IWUSR)) == -1) {
		g_error(_("Unable to open %s for writing."), output);
		close (fdo);
		return -1;
	}

	while (1) {
		if ((len = read (fd, data, 4)) != 4) {
			if (len == 0) {
				break;
			} else {
				g_error(_("Reading firmware header chunk "
					  "failed"));
				goto end;
			}
		}
		len = (data[0] << 8) | data[1];
		req = (data[2] << 8) | data[3];
		if (len == 0)
			continue;
		else if (len < 0 || len >= 1024) {
			g_error(_("Invalid firmware data_length %d, "
				  "load aborted"), len);
			goto end;
		} else if (read (fd, rdata, len) != len) {
			g_error(_("Error reading firmware data"));
			goto end;
		}

		write_chunk(fdo, len, req, rdata);
	}

	/* Intel HEX end of firmware */
	const gchar *end = ":00000001FF\n";
	write(fdo, end, strlen(end));

	ret = 0;
 end:
	close(fd);
	fchmod(fdo, 0644);
	close(fdo);

	return 0;
}

int main (int argc, char *argv[])
{
	GOptionContext *context;
	GError *error = NULL;
	int status = 0;
	
#if ENABLE_NLS
	/* ??????? */
	setlocale(LC_ALL, "");
	bindtextdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);
#endif

	gchar *appname = _("built-in iSight firmware exporter");

	g_set_application_name(appname);
	g_set_prgname("ift-export");

	/* translators: this is the first --help output line */
	context = g_option_context_new(g_strdup_printf("- %s",
						       appname));
	g_option_context_add_main_entries(context, entries, GETTEXT_PACKAGE);
	if (!g_option_context_parse(context, &argc, &argv, &error)) {
		fprintf(stderr, _("Error: %s\n"),
			error->message); /* comment traduire ? */
		fprintf(stderr, g_option_context_get_help(context,
							  FALSE, NULL));
	}

	if (g_access(input, R_OK)) {
		g_error(_("Unable to read firmware %s."), input);
	}

	g_option_context_free(context);

	/* exoirt */
	status = export(input, output);
	if (!status)
		g_message(_("Firmware exported successfully in %s"), output);

 	return status;
}


src/extract.c
/*
 * Apple built-in iSight firmware extractor
 *
 * Copyright © 2006 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright © 2007 Étienne Bersac <bersace@gmail.com>
 *
 *
 * This software extract firmware from any Mac OS X iSight driver
 * (named AppleUSBVideoSupport).
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA
 *
 *
 *
 * 
 * - First version had hardcoded offset, with a SHA1 sum check
 *   (Ronald)
 *   
 * - Second version, multiple sha1sum were handled with their offset
 *   (Étienne)
 *   
 * - Third version, we search for a preambule in driver and extract
 *   the firmware (Étienne, idea from Jeroen van der Vegt)
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <fcntl.h>
#include <unistd.h>
#include <gcrypt.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gstdio.h>

/*
 * The vanilla firmware have bugs in USB Interface descriptor
 * values. We override buggy value with correct one. Those value are
 * triplet, we replace 0xFF 0xFF 0xFF by specific value.
 */
struct patch {
	const char* desc;
	long offset;
	unsigned char value[3];
};

/* From linux-2.6:linux/usb/ch9.J */
#define	USB_CLASS_VIDEO			0x0E
#define USB_CLASS_MISC			0xEF
#define USB_PROTOCOL_IFACE_ASSOC	0x01

/* From linux-uvc: uvcvideo.h */
#define SC_VIDEOCONTROL                 0x01
#define SC_VIDEOSTREAMING               0x02
#define SC_VIDEO_INTERFACE_COLLECTION   0x03
#define PC_PROTOCOL_UNDEFINED           0x00

static struct patch patches[] = {
	{ N_("Fix device descriptor"),
	  .offset	= 0x2352,
	  .value	= { USB_CLASS_MISC,
			    0x02,
			    USB_PROTOCOL_IFACE_ASSOC }
	},
	{ N_("Fix interface assocation descriptor"),
	  .offset	= 0x2373,
	  .value	= { 0x08, 0x0b, 0x00 }
	},
	{ N_("Fix video interface collection"),
	  .offset	= 0x2377,
	  .value	= { USB_CLASS_VIDEO,
			    SC_VIDEO_INTERFACE_COLLECTION,
			    PC_PROTOCOL_UNDEFINED }
	},
	{ N_("Fix video streaming device qualifier"),
	  .offset	= 0x2364,
	  .value	= { USB_CLASS_VIDEO,
			    SC_VIDEOSTREAMING,
			    PC_PROTOCOL_UNDEFINED }
	},
	{ N_("Fix video control interface descriptor"),
	  .offset	= 0x2380,
	  .value	= { USB_CLASS_VIDEO,
			    SC_VIDEOCONTROL,
			    PC_PROTOCOL_UNDEFINED }
	},
	{ N_("Fix video streaming interface descriptor"),
	  .offset	= 0x23C6,
	  .value	= { USB_CLASS_VIDEO,
			    SC_VIDEOSTREAMING,
			    PC_PROTOCOL_UNDEFINED }
	},
};

/* OPTIONS */
static gchar *driver_filename	= NULL;
static gchar *firmware_dir	= "/lib/firmware";
static gchar *firmware_filename = "isight.fw";

static GOptionEntry entries[] = {
	{ "apple-driver", 'a',
	  0,
	  G_OPTION_ARG_FILENAME, &driver_filename,
	  N_("Path to the AppleUSBVideoSupport driver file."),
	  NULL
	},
	{ "output-directory", 'd',
	  0,
	  G_OPTION_ARG_FILENAME, &firmware_dir,
	  N_("Output firmware directory"),
	  "/lib/firmware"
	},
	{ "output-filename", 'f',
	  0,
	  G_OPTION_ARG_FILENAME, &firmware_filename,
	  N_("Name of the output firmware file"),
	  "isight.fw"
	},
	{ NULL }
};


/* IMPLEMENTATION */

/*
 * To find the offset in the driver file, use a hexadecimal editor
 * (e.g. ghex) and search for "USBSend". You will find data like
 * this :
 *
 * "start USBSend #4 returned 0x%x.....X.....4.~.........%\"
 *                                                  ^
 *
 * Each period means 0x00. The firmware start at the ^ sign
 * position. Read the offset and store in in this array. The tool must
 * output the sha1sum of the file, use this to fill the .sha1sum
 * field.
 */
static glong
seek_offset(const gchar* filename)
{
	gint fd;
	glong pos;
	/* This is the preambule of the firmware in the driver file */
	const guchar needle[8] = {0xC2, 0x34, 0x06, 0x7E,
				  0x05, 0x00, 0x00, 0x00};
	guint read_len, chunk_len = 8;
	guint offset = 8;
	guchar* chunk = g_new0(guchar, chunk_len);

	if ((fd = g_open (filename, O_RDONLY)) == -1) {
		g_error(_("Unable to open %s."), filename);
		pos = -1;
		goto end;
	}

	for (pos = 0x1000; pos < 0xD000; pos++) {
		if (lseek(fd, pos, SEEK_SET) != pos) {
			g_error(_("Failed to seek %s at position %x."),
				filename, (guint) pos);
			break;
		}

		if ((read_len = read(fd, chunk, chunk_len)) != chunk_len) {
			g_error(_("Failed to read %s at position %x."),
				filename, (guint) pos);
			break;
		}

		if (memcmp(needle, chunk, chunk_len) == 0) {
			pos+= offset;
			g_message(_("Found firmware signature at offset 0x%X."),
				  (guint) pos);
			goto end_fd;
		}
	}
	pos = -1;


 end_fd:
	close(fd);
 end:
	g_free(chunk);
	return pos;
}


/* extract firmware from filename at offset to output, checking the
   format. */
static int
extract(char *filename,
	long offset,
	char*output)
{
	int fd, fdo, len, req, i, ret = -1;
	unsigned char data[4], rdata[1024];

	if ((fd = g_open (filename, O_RDONLY)) == -1) {
		g_error(_("Unable to open %s."), filename);
		return -1;
	} else if (lseek (fd, offset, SEEK_SET) != offset) {
		g_error(_("Failed to seek %s at position %x."),
			filename, (guint) offset);
		close (fd);
		return -1;
	}

	if ((fdo = g_open(output, O_WRONLY | O_CREAT | O_TRUNC,
			  S_IRUSR | S_IWUSR)) == -1) {
		g_error(_("Unable to open %s for writing."), output);
		close (fdo);
		return -1;
	}

	while (1) {
		if ((len = read (fd, data, 4)) != 4) {
			if (len == 0) {
				g_error(_("Unexpected EOS - corrupt driver?"));
				goto end;
			} else {
				g_error(_("Reading firmware header chunk "
					  "failed"));
				goto end;
			}
		}
		len = (data[0] << 8) | data[1];
		req = (data[2] << 8) | data[3];
		if (len == 0x8001) {
			break;	/* success */
		} else if (len == 0)
			continue;
		else if (len < 0 || len >= 1024) {
			g_error(_("Invalid firmware data_length %d, "
				  "load aborted"), len);
			goto end;
		} else if (read (fd, rdata, len) != len) {
			g_error(_("Error reading firmware data"));
			goto end;
		}

		/* write to file */
		i = write(fdo, data, 4);
		i = write(fdo, rdata, len);
	}

	ret = 0;
 end:
	close(fd);
	fchmod(fdo, 0644);
	close(fdo);

	return 0;
}

static gint
firmware_version(int fd, guchar* version)
{
	guint read_len;

	if (lseek(fd, 4, SEEK_SET) == -1)
		g_error(_("Failed to find firmware version, unable to patch it"));
		
	if ((read_len = read(fd, version, 3)) != 3)
		g_error(_("Failed to read firmware version, unable to patch it"));

	g_message(_("Firmware version %d.%d.%d (0x%02X.0x%02X.0x%02X)"),
		  version[0], version[1], version[2],
		  version[0], version[1], version[2]);

	return 0;
}

static int
patch(char *filename)
{
	int fd, i, n, ret = 0;
	guchar version[3];

	if ((fd = g_open(filename, O_RDWR)) == -1) {
		g_error(_("Error while opening firmware file for patching."));
		return -1;
	}

	firmware_version(fd, version);

	n = G_N_ELEMENTS(patches);
	for (i = 0; i < n; i++) {
		g_message(_("Apply patch %i : %s"),
			  i, gettext(patches[i].desc));

		if (lseek(fd, patches[i].offset, SEEK_SET) == -1) {
			g_error(_("Unable to patch the firmware."));
			ret = -1;
			goto end;
		}

		if (write(fd, &patches[i].value, 3) == 0) {
			g_error(_("Failed to write patched value !"));
			ret = -1;
			goto end;
		}
	}

 end:
	close(fd);
	return ret;
}


int main (int argc, char *argv[])
{
	GOptionContext *context;
	GError *error = NULL;
	int status = 0;
	gchar *pathname;
	glong offset;

#if ENABLE_NLS					\
	/* ??????? */
	setlocale(LC_ALL, "");
	bindtextdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);
#endif

	gchar *appname = _("built-in iSight firmware extractor "
			   "and patcher");

	g_set_application_name(appname);
	g_set_prgname("ift-extract");

	/* translators: this is the first --help output line */
	context = g_option_context_new(g_strdup_printf("- %s",
						       appname));
	/* translators: this is shown at the end of --help output */
	gchar *desc = g_strdup_printf(_("AppleUSBVideoSupport driver is usualy "					"found in /System/Library/Extensions/"
					"IOUSBFamily.kext/Contents/PlugIns/"
					"AppleUSBVideoSupport.kext/Contents/"
					"MacOS/ in your Mac OS X root volume."
					"\n\n"
					"If you have a non-working Apple "
					"driver file, please send it to %s with"
					"machine description and OS X "
					"version."),
				      PACKAGE_BUGREPORT);
	g_option_context_set_description(context, desc);
	g_option_context_add_main_entries(context, entries, GETTEXT_PACKAGE);
	if (!g_option_context_parse(context, &argc, &argv, &error)) {
		fprintf(stderr, _("Error: %s\n"),
			error->message); /* howto translate? */
		fprintf(stderr, g_option_context_get_help(context,
							  FALSE, NULL));
		return 1;
	}

	if (driver_filename == NULL) {
	  fprintf(stderr, _("Option -a is mandatory. See --help.\n"));
	  return 1;
	}

	if (g_access(driver_filename, R_OK | F_OK)) {
		g_error(_("Unable to read driver %s."), driver_filename);
		return 1;
	}

	pathname = g_build_filename(firmware_dir, firmware_filename, NULL);
	g_option_context_free(context);

	/* search */
	offset = seek_offset(driver_filename);

	if (offset <= 0)
		g_error(_("Unable to find firmware in the file."));

	/* extract */
	status = extract(driver_filename, offset, pathname);
	if (!status)
		g_message(_("Firmware extracted successfully in %s"), pathname);

	/* patch */
	status = patch(pathname);
	if (!status)
		g_message(_("Firmware patched successfully"));
	else
		g_error(_("Failed to apply patches to %s"), pathname);

 	return status;
}
